wire-gradle-plugin/src/main/kotlin/com/squareup/wire/gradle/WireExtension.kt

```diff
@@ -0,0 +1,134 @@
+/*
+ * Copyright (C) 2019 Square, Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package com.squareup.wire.gradle
+
+import groovy.lang.Closure
+import org.gradle.api.Action
+import org.gradle.api.Project
+import org.gradle.api.file.SourceDirectorySet
+import org.gradle.api.internal.file.SourceDirectorySetFactory
+import org.gradle.api.tasks.Input
+import org.gradle.api.tasks.InputFiles
+import org.gradle.api.tasks.Optional
+import org.gradle.util.ConfigureUtil
+import java.util.Arrays
+import javax.inject.Inject
+
+open class WireExtension(
+  project: Project,
+  private val sourceDirectorySetFactory: SourceDirectorySetFactory
+) {
+  private val objectFactory = project.objects
+
+  internal val sourcePaths = mutableSetOf<String>()
+  internal val protoPaths = mutableSetOf<String>()
+  internal val sourceTrees = mutableSetOf<SourceDirectorySet>()
+  internal val protoTrees = mutableSetOf<SourceDirectorySet>()
+
+  @get:Input
+  @get:Optional
+  var roots: Array<String>? = null
+
+  @get:Input
+  @get:Optional
+  var prunes: Array<String>? = null
+
+  @get:Input
+  @get:Optional
+  var rules: String? = null
+
+  @get:Input
+  @get:Optional
+  var javaTarget: JavaTarget? = null
+
+  @get:Input
+  @get:Optional
+  var kotlinTarget: KotlinTarget? = null
+
+  @InputFiles
+  @Optional
+  fun getSourcePaths() = sourcePaths.toSet()
+
+  @InputFiles
+  @Optional
+  fun getSourceTrees() = sourceTrees.toSet()
+
+  /**
+   * Source paths for local jars and directories, as well as remote binary dependencies
+   */
+  fun sourcePath(vararg sourcePaths: String) {
+    this.sourcePaths.addAll(Arrays.asList(*sourcePaths))
+  }
+
+  /**
+   * Source paths for local file trees, backed by a [org.gradle.api.file.SourceDirectorySet]
+   * Must provide at least a [org.gradle.api.file.SourceDirectorySet.srcDir]
+   */
+  fun sourcePath(closure: Closure<SourceDirectorySet>) {
+    val sourceTree = sourceDirectorySetFactory.create("source-tree", "Source path tree")
+    sourceTree.filter.include("**/*.proto")
+    ConfigureUtil.configure<SourceDirectorySet>(closure, sourceTree)
+    sourceTrees.add(sourceTree)
+  }
+
+  @InputFiles
+  @Optional
+  fun getProtoPaths(): Set<String> {
+    return protoPaths
+  }
+
+  @InputFiles
+  @Optional
+  fun getProtoTrees(): Set<SourceDirectorySet> {
+    return protoTrees
+  }
+
+  fun protoPath(vararg protoPaths: String) {
```

Comment:
I like KDocs on `sourcePath()`, can we include them on all methods/properties in this class? Fine as a follow-up

Reply:
Added a few more KDocs to the other properties.